src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -413,12 +420,109 @@ void FancyZones::MoveWindowIntoZone(HWND window, winrt::com_ptr<IZoneWindow> zon
+    int top = destMonitorRect.top + (windowRect.top - originMonitorRect.top);
+    int W = RectWidth(windowRect);
+    int H = RectHeight(windowRect);
+
+    if ((left < destMonitorRect.left) || (left + W > destMonitorRect.right))
+    {
+        // Set left window border to left border of screen (add padding). Resize window width if needed.
+        left = destMonitorRect.left + LEFT_TOP_PADDING;
+        W = min(W, RectWidth(destMonitorRect) - LEFT_TOP_PADDING);
+    }
+    if ((top < destMonitorRect.top) || (top + H > destMonitorRect.bottom))
+    {
+        // Set top window border to top border of screen (add padding). Resize window height if needed.
+        top = destMonitorRect.top + LEFT_TOP_PADDING;
+        H = min(H, RectHeight(destMonitorRect) - LEFT_TOP_PADDING);
+    }
+
+    return { .left   = left,
+             .top    = top,
+             .right  = left + W,
+             .bottom = top + H };
+}
+
+void OpenWindowOnActiveMonitor(HWND window, HMONITOR monitor) noexcept
+{
+    // By default Windows opens new window on primary monitor.
+    // Try to preserve window width and height, adjust top-left corner if needed.
+    HMONITOR origin = MonitorFromWindow(window, MONITOR_DEFAULTTOPRIMARY);
+    if (origin == monitor)
+    {
+        // Certain applications by design open in last known position, regardless of FancyZones.
+        // If that position is on currently active monitor, skip custom positioning.
+        return;
+    }
+
+    WINDOWPLACEMENT placement{};
+    if (GetWindowPlacement(window, &placement))
+    {
+        MONITORINFOEX originMi;
+        originMi.cbSize = sizeof(originMi);
+        if (GetMonitorInfo(origin, &originMi))
+        {
+            MONITORINFOEX destMi;
+            destMi.cbSize = sizeof(destMi);
+            if (GetMonitorInfo(monitor, &destMi))
+            {
+                RECT newPosition = FitOnScreen(placement.rcNormalPosition, originMi.rcWork, destMi.rcWork);
+                SizeWindowToRect(window, newPosition);
+
+                // Certain applications could perform their own custom positioning afterwards, so check if window
+                // is in expected position, and if not, try to move it there again.
+                for (int i = 0; i < CUSTOM_POSITIONING_MAX_RETRIES; ++i)
+                {
+                    std::this_thread::sleep_for(std::chrono::milliseconds(CUSTOM_POSITIONING_WAIT_TIME));
+                    if (WindowInExpectedPosition(window, newPosition))
+                    {
+                        break;
+                    }
+                    SizeWindowToRect(window, newPosition);
+                }
+            }
+        }
+    }
+}
+
 // IFancyZonesCallback
 IFACEMETHODIMP_(void)
 FancyZones::WindowCreated(HWND window) noexcept
 {
     std::shared_lock readLock(m_lock);
-    if (m_settings->GetSettings()->appLastZone_moveWindows && ShouldProcessNewWindow(window))
+    if (ShouldProcessNewWindow(window))

```

Comment:
We can optimize the code here and first check if we really need to proceed.
`m_settings->GetSettings()->appLastZone_moveWindows` and `m_settings->GetSettings()->openWindowOnActiveMonitor` are low cost checks that should be performed before `ShouldProcessNewWindow`

Reply:
Added checks for these two before `ShouldProcessNewWindow` call.